# Volatility Modeling Using Quantitative Methods

In [1]:
import pandas as pd
import numpy as np
import yfinance as yf
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import warnings
warnings.filterwarnings("ignore")

# TRADITIONAL RISK MODELS

### ARCH MODEL

In [2]:
from arch import arch_model
tickers=['JPM']
df=yf.download(tickers, start='2022-01-01', end='2025-01-01')
df['returns']=df['Close'].pct_change()*100 # Daily % Return
returns=df['returns'].dropna()

[*********************100%***********************]  1 of 1 completed


In [3]:
# Call the Model
model=arch_model(returns, vol='ARCH', p=1) 
# Train The Model
results=model.fit(disp='off')
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                      Constant Mean - ARCH Model Results                      
==============================================================================
Dep. Variable:                returns   R-squared:                       0.000
Mean Model:             Constant Mean   Adj. R-squared:                  0.000
Vol Model:                       ARCH   Log-Likelihood:               -1400.95
Distribution:                  Normal   AIC:                           2807.90
Method:            Maximum Likelihood   BIC:                           2821.76
                                        No. Observations:                  752
Date:                Tue, Mar 31 2026   Df Residuals:                      751
Time:                        10:33:22   Df Model:                            1
                                 Mean Model                                
===========================================================================
                 coef    std err          t      P>|t|     95.0% Conf. Int.
---------------------------------------------------------------------------
mu             0.0978  5.518e-02      1.772  7.646e-02 [-1.039e-02,  0.206]
                              Volatility Model                             
===========================================================================
                 coef    std err          t      P>|t|     95.0% Conf. Int.
---------------------------------------------------------------------------
omega          2.1877      0.332      6.584  4.591e-11    [  1.536,  2.839]
alpha[1]       0.1224  7.467e-02      1.639      0.101 [-2.394e-02,  0.269]
===========================================================================

Covariance estimator: robust
"""

Analysis of the ARCH Model
1. mu => 0.0978 : The model estimates that the average daily return is 0.1260
2. omega => 2.1877: The long run base level of variance
3. alpha[1] => 0.1224: How much yesterday's squared shock impacts today's variance

In [4]:
# Forecast for next 5 days
forecast=results.forecast(horizon=5)
predicted_variance=forecast.variance
predicted_volatility=np.sqrt(predicted_variance)
predicted_volatility=[1.290791,	1.538572,	1.631696	,1.669358	,1.684967]
predicted_avg_volatility=sum(predicted_volatility)/len(predicted_volatility)
print(predicted_avg_volatility)

1.5630768


In [5]:
# Actual Volatility
start_date=pd.to_datetime('2024-12-31')
end_date=pd.to_datetime('2025-01-09')

real_df=yf.download('JPM', start=start_date, end=end_date)
real_df['returns']=real_df['Close'].pct_change()*100
real_df=real_df.dropna()

realized_vol=real_df['returns'].std()*np.sqrt(5) # As we are predicting for 5 days
print(realized_vol)

[*********************100%***********************]  1 of 1 completed

1.691954204929747


In [6]:
print("ARCH Model Predicted Volatility: ", predicted_avg_volatility)
print("ARCH Model Actual Volatility: ", realized_vol)

ARCH Model Predicted Volatility:  1.5630768
ARCH Model Actual Volatility:  1.691954204929747


# GARCH Model

In [7]:
from arch import arch_model
tickers=['JPM']
df=yf.download(tickers, start='2022-01-01', end='2025-01-01')
df['returns']=df['Close'].pct_change()*100 # Daily % Return
returns=df['returns'].dropna()

[*********************100%***********************]  1 of 1 completed


In [8]:
# Call the Model
model=arch_model(returns, vol='GARCH', p=1, q=1) 

# Train The Model
results=model.fit(disp='off')

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                     Constant Mean - GARCH Model Results                      
==============================================================================
Dep. Variable:                returns   R-squared:                       0.000
Mean Model:             Constant Mean   Adj. R-squared:                  0.000
Vol Model:                      GARCH   Log-Likelihood:               -1387.47
Distribution:                  Normal   AIC:                           2782.94
Method:            Maximum Likelihood   BIC:                           2801.43
                                        No. Observations:                  752
Date:                Tue, Mar 31 2026   Df Residuals:                      751
Time:                        10:34:43   Df Model:                            1
                                Mean Model                                
==========================================================================
                 coef    std err          t      P>|t|    95.0% Conf. Int.
--------------------------------------------------------------------------
mu             0.1146  5.754e-02      1.991  4.648e-02 [1.786e-03,  0.227]
                               Volatility Model                              
=============================================================================
                 coef    std err          t      P>|t|       95.0% Conf. Int.
-----------------------------------------------------------------------------
omega          0.0220  1.702e-02      1.294      0.196 [-1.134e-02,5.537e-02]
alpha[1]       0.0165  1.112e-02      1.481      0.139 [-5.326e-03,3.827e-02]
beta[1]        0.9737  1.217e-02     79.976      0.000      [  0.950,  0.998]
=============================================================================

Covariance estimator: robust
"""

Analysis of the GARCH Model
1. mu => 0.1146 : THe model estimates that the average daily return is 0.1260
2. omega => 0.0220: The long run base level of variance
3. alpha[1] => 0.0165: How much yesterday's squared shock impacts today's variance
4. beta[1]=> 0.9737: How much yesterday's variance impacts today's variance

In [9]:
# Forecast for next 5 days
forecast=results.forecast(horizon=5)
predicted_variance=forecast.variance
predicted_volatility=np.sqrt(predicted_variance)
predicted_volatility=[1.623482	,1.622251,	1.621031,	1.619823,	1.618625]
predicted_avg_volatility=sum(predicted_volatility)/len(predicted_volatility)
print(predicted_avg_volatility)

1.6210423999999999


In [10]:
# Actual Volatility
start_date=pd.to_datetime('2024-12-31')
end_date=pd.to_datetime('2025-01-09')

real_df=yf.download('JPM', start=start_date, end=end_date)
real_df['returns']=real_df['Close'].pct_change()*100
real_df=real_df.dropna()

realized_vol=real_df['returns'].std()*np.sqrt(5) # As we are predicting for 5 days
print(realized_vol)

[*********************100%***********************]  1 of 1 completed

1.691954204929747


In [11]:
print("GARCH Model Predicted Volatility: ", predicted_avg_volatility)
print("GARCH Model Actual Volatility: ", realized_vol)

GARCH Model Predicted Volatility:  1.6210423999999999
GARCH Model Actual Volatility:  1.691954204929747


# EWMA Model

In [12]:
# Download the data
tickers=['JPM']
df=yf.download(tickers, start='2024-01-01', end='2025-01-01')
df['returns']=df['Close'].pct_change()*100 # Daily % Return
df=df.dropna()

[*********************100%***********************]  1 of 1 completed


In [13]:
# Calculate EWMA
lamda= 0.94
ewma_var=[]
var_t=df['returns'].var()

for ret in df['returns']:
    variance_tplus1=lamda*var_t+(1-lamda)*(ret**2)
    ewma_var.append(variance_tplus1)

df['ewma_vol']=np.sqrt(ewma_var)


In [14]:
# Predicted Volatility (EWMA model can only be used for 1 day prediction)
latest_daily_vol=df['ewma_vol'].iloc[-1]
print(latest_daily_vol)

1.4369557946028217


In [15]:
start_date=pd.to_datetime('2024-12-31')
end_date=pd.to_datetime('2025-01-03')
real_df=yf.download('JPM', start=start_date, end=end_date)
real_df['returns']=real_df['Close'].pct_change()*100
real_df=real_df.dropna()

realized_vol=real_df['returns'] 
print(realized_vol)

[*********************100%***********************]  1 of 1 completed

Date
2025-01-02    0.120967
Name: returns, dtype: float64


In [16]:
print("EWMA Model Predicted Volatility: ", latest_daily_vol)
print("EWMA Model Actual Volatility: ", realized_vol)

EWMA Model Predicted Volatility:  1.4369557946028217
EWMA Model Actual Volatility:  Date
2025-01-02    0.120967
Name: returns, dtype: float64
